cmd/kubeadm/app/phases/addons/addons_test.go

```diff
@@ -86,3 +92,69 @@ func TestCompileManifests(t *testing.T) {
 		}
 	}
 }
+
+// This unit test exists because tolerations weren't decoded correctly by kuberuntime.DecodeInto
+// In case that regression ever happens again, this unit test will catch it
+func TestTolerationsPreserved(t *testing.T) {
+	var tests = []struct {
+		manifest string
+		data     interface{}
+		f        func(data []byte) []v1.Toleration
+		expected []v1.Toleration
+	}{
+		{
+			manifest: KubeProxyDaemonSet,
+			data: struct{ Image, ClusterCIDR, MasterTaintKey string }{
+				Image:          "foo",
+				ClusterCIDR:    "- foo",
+				MasterTaintKey: kubeadmconstants.LabelNodeRoleMaster,
+			},
+			f: func(data []byte) []v1.Toleration {
+				kubeproxyDaemonSet := &extensions.DaemonSet{}
+				if err := kuberuntime.DecodeInto(api.Codecs.UniversalDecoder(), data, kubeproxyDaemonSet); err != nil {
```

Comment:
Oh, well, I didn't even realize this was `k8s.io/kubernetes/pkg/api` while we are using ` k8s.io/client-go/pkg/api{,/v1}` everywhere else...

> Maybe the UniversalDecoder() doesn't bother to infer the type of kubeproxyDaemonSet

That was my understanding though, and it has always kind of worked.
@mikedanese wrote the initial code for this and may be able to comment more in detail

cc @deads2k @sttts as well ^